### DISCUSSION TITLE
Support hiding some JRE packages from the plugin

### DISCUSSION THREAD

**Discussion by testforstephen** [eclipse-equinox/equinox#697] [Q&A] [answered]
**Use Case:**
I'm working on a project that leverages javac as both a parser and compiler to provide Java language support. Previously, our tool relied on the javac bundled with the installed JDK, requiring users to install the latest JDK to get the new feature support. To eliminate this dependency, we are trying to embed a standalone javac JAR in our plugin to provide these capabilities independently of the system JDK.

However, a challenge arises because the embedded javac JAR contains classes that overlap with those in JRE modules, such as java.compiler and jdk.compiler. In practice, the JRE's classes take precedence, so the plugin classloader prioritizes the JRE versions of these classes over those in our embedded javac.

I'm looking for guidance on whether the OSGi platform provides a way to hide specific packages on the classpath from the plugin and allow it to use its own versions of overlapping classes.

**Comment by laeubi** [comment]
@testforstephen due to technical limitations you cannot "replace" any package from the `java.*` namespace, but if you embedd the jar and simply don't import the package otherwise you will get it from your embedded jar (I assume its on the Bundle-Classpath...).

**Comment by merks** [comment]
JustJ provides JREs that include javac:

https://eclipse.dev/justj/?page=documentation

![image](https://github.com/user-attachments/assets/ae48e9de-f177-4790-ac34-f92850b03b4d)

I wonder if you can't just reuse that.

As @laeubi suggests, I don't think your use case can be supported...

**Comment by testforstephen** [comment]
@merks @laeubi thanks for the response. Our project (JDT language server) uses javac API directly rather than invoking the javac.exe. The namespaces provided by the javac API are in the `javax.*` and `com.sun.*` packages. Currently, I’ve embedded a javac.jar in my bundle and set it on the `Bundle-ClassPath` in the bundle's MANIFEST.MF. Eclipse compiles this setup without issues, but at runtime, the bundle continues to use the javac API from the JRE instead of the embedded version. 

If there is no existing solution for it yet, would it be possible to use a custom classloader to prioritize loading our local version over the JRE’s? Or could we extend the MANIFEST.MF to support an attribute like `Hide-Classpath-Package` to allow hiding packages on the classpath from the bundle?

**Comment by laeubi** [comment]
`javax.*` and `com.sun.*` should be fine, you should check the following things:

1. There is no `Import-Package` for these packages you want to use from the jar, otherwise they might get priority
2. The jar is included in the binary build (e.g via `build.properties`)
3. It could be that you need to disable delegate classlaoding with `osgi.compatibility.bootdelegation=false` see [here](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/misc/runtime-options.html?cp=2_1_6_0) but this depends on your deployment configuration.
4. If it is allowed to embed /ship `javac.jar` from a legal perspective

**Comment by merks** [comment]
I think you should check out what the JDT project is doing to enable using javac for building the abstract syntax tree.   Maybe they already solved this problem?

**Comment by testforstephen** [comment]
> javax.* and com.sun.* should be fine, you should check the following things:
> 1. There is no Import-Package for these packages you want to use from the jar, otherwise they might get priority
> 2. The jar is included in the binary build (e.g via build.properties)
> 3. It could be that you need to disable delegate classlaoding with osgi.compatibility.bootdelegation=false see [here](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/misc/runtime-options.html?cp=2_1_6_0) but this depends on your deployment configuration.
>4. If it is allowed to embed /ship javac.jar from a legal perspective

@laeubi I have configured the local javac.jar path in both `Bundle-ClassPath` of MANIFEST.MF and `bin.includes` of `build.properties`. And in the `Import-Package` attribute, I didn't configure `javax.*` and `com.sun.*` packages. In Eclipse, when I navigate to javac API references in my code, it correctly links to the local javac.jar.

However, during debugging the bundle, the Java runtime prioritizes the system module when finding javac api classes. I use the Java command line parameters `-verbose:class --show-module-resolution` to view the class/module loading logs, I can see that the runtime loads the javac classes from the JRE’s modules (e.g. `java.compiler`, `jdk.compiler`) instead of the local javac.jar. This indicates the bundle classloader prioritizes JDK classes over local classes.

I will take a look at `osgi.compatibility.bootdelegation=false` to see if it helps.

> I think you should check out what the JDT project is doing to enable using javac for building the abstract syntax tree. Maybe they already solved this problem?

@merks JDT is using ECJ (Eclipse Compiler for Java) to build AST, the thing we're doing is to make JDT to support an alternative parser/compiler (e.g. javac).

**Comment by merks** [comment]
I suggest you revisit what jdt is doing now because it’s doing exactly what you just stated.

**Comment by akurtakov** [comment]
> I suggest you revisit what jdt is doing now because it’s doing exactly what you just stated.

@merks What you speak about is being done in subproject of jdt - jdt.ls (https://projects.eclipse.org/projects/eclipse.jdt.ls) and is not (yet!) part of main jdt. Also @testforstephen is one of the main drivers of that works and his request is in order to make this work easier.
I hope this clears the confusion.

**Comment by laeubi** [comment]
@testforstephen Its a bit hard to guess what happens, maybe you can share a small example?

In general the class-loading works like described here:

https://docs.osgi.org/specification/osgi.core/8.0.0/framework.module.html#i3174728

and as mentioned above, the only package that can't be overridden is `java.*` so there must be something fishy in your setup or interpretation of results.

**Comment by tjwatson** [comment, marked as answer]
I agree with @laeubi there must be something strange with your bundle.  Like requiring the system bundle or `org.eclipse.osgi`.  Keep in mind that using require-bundle here will cause issues.  For example, using `Require-Bundle` for `org.eclipse.core.runtime` will pull in all of `org.eclipse.osgi` which in-turn exposes you to all the JDK packages because this horrible re-export:

https://github.com/eclipse-platform/eclipse.platform/blob/6dd67323474b1b49d541bcd1e4ea8007ab2a36a4/runtime/bundles/org.eclipse.core.runtime/META-INF/MANIFEST.MF#L12C17-L12C87

Something like that will prevent you from loading the packages from your bundle's class path.

**Comment by merks** [comment]
1.  I expect it would be helpful to see concretely **exactly** what is currently being tried because you'll note that most of the answers are kind of theoretical based on hypotheses about what's actually being done.
2. This issue appears to be more like a GitHub discussion than an actionable GitHub issue, especially given its asking for guidance and it provides no concrete way of reproducing "a/the" problem.

**Comment by testforstephen** [comment]
@merks @laeubi @tjwatson @akurtakov thank you all for the suggestions.

Here are the reproducing steps:
- Open the sample project in Eclipse. [org.eclipse.jdt.javac.experiment.zip](https://github.com/user-attachments/files/17763318/org.eclipse.jdt.javac.experiment.zip)
- Launch it as an Eclipse application. You can add `-verbose:class` to vmArgs to enable the class load logs.
- In the debuggee Eclipse window, click the menu `Sample Menu -> Sample Command` that will trigger the target javac api code.

As a result, it throws errors because javac api is loaded from `jrt:/java.compiler` rather than my local nb-javac.jar. You can view the class load logs in the original Eclipse Console viewlet, and search the keywords such as "javax.tools", "com.sun".

A video to show what I observed.

https://github.com/user-attachments/assets/43697f9e-b4a9-43cc-a01b-a371ab2a9c24

**Reply by laeubi** [reply]
@testforstephen if you loo at the manifest and recap what @tjwatson wrote you will see that exactly this case here applies:

>  For example, using Require-Bundle for org.eclipse.core.runtime will pull in all of org.eclipse.osgi which in-turn exposes you to all the JDK packages because this horrible re-export

![grafik](https://github.com/user-attachments/assets/29c4cc77-1085-4ad6-baa3-e93b22606bfc)

I removed the both required bundles from your manifest, used a plain OSGi activator instead of an UI handler to trigger your test-code and if I then start the bundle the code is successfully executed.

**Reply by testforstephen** [reply]
> I removed the both required bundles from your manifest, used a plain OSGi activator instead of an UI handler to trigger your test-code and if I then start the bundle the code is successfully executed.

Not quite understand how to make it work. Do you mean my bundle can't use `org.eclipse.core.runtime` and `org.eclipse.osgi` as dependnecies? I'm a little confused here. Can you explain more about the reason? thanks.

**Reply by laeubi** [reply]
> Not quite understand how to make it work. Do you mean my bundle can't use `org.eclipse.core.runtime` and `org.eclipse.osgi` as dependencies?

Basically, yes. The reason is that `org.eclipse.core.runtime` reexports all packages of `org.eclipse.osgi`  and these import all system packages, so your bundle indirectly imports the packages you want to hide.

I also don't see a reason why a language server would ever require these here...

**Reply by mickaelistria** [reply]
> The reason is that org.eclipse.core.runtime reexports all packages of org.eclipse.osgi and these import all system packages

OK, So one issue in our particular case is that we get this `Require-Bundle: org.eclipse.core.runtime` from the JDT fragment. @testforstephen we should investigate replacing this in JDT Core by some Import-Package then.

**Reply by laeubi** [reply]
> OK, So one issue in our particular case is that we get this Require-Bundle: org.eclipse.core.runtime from the JDT fragment. @testforstephen we should investigate replacing this in JDT Core by some Import-Package then.

Please be aware that Import-Package might not work for JDT because we have split packages. Anyways often another bundle (e.g. equinox.common) can be required but this all depends heavily on why it is required (for what classes) at all.

Also it might be good to simply do the javac stuff in an own bundle (so it has an own classloader) and only communicate with a dedicated service interface.

**Reply by testforstephen** [reply]
> > The reason is that org.eclipse.core.runtime reexports all packages of org.eclipse.osgi and these import all system packages
> 
> OK, So one issue in our particular case is that we get this `Require-Bundle: org.eclipse.core.runtime` from the JDT fragment. @testforstephen we should investigate replacing this in JDT Core by some Import-Package then.

@mickaelistria I understand the reason why we make javac bundle to be a fragment of JDT Core because our javac DOM parser accesses some package-private classes from JDT Core. Do you think how possible to get rid of these package-private usages in javac bundle?

**Reply by mickaelistria** [reply]
> Do you think how possible to get rid of these package-private usages in javac bundle?

I think it would be a lot of work, and it's IMO safer for JDT to not open further packages. It's IMO simpler to evaluate replacing the `Require-Bundle: org.eclipse.core.runtime` in JDT at this stage.
Other approaches such as making JDT open more packages, or creating some intermediary DOM Binding model as API between JDT Core and JDT.Javac and so on would take much more time with a much higher risk of failure. Trying to change the Require-Bundle seems like a low-hanging fruit compared to any other approach I can think of.

**Reply by jukzi** [reply]
> Also it might be good to simply do the javac stuff in an own bundle (so it has an own classloader) and only communicate with a dedicated service interface.

+1

**Comment by testforstephen** [comment]
@tjwatson @laeubi Thank you for the suggestions. Replacing `Require-Bundle` with `Import-Package` for `org.eclipse.core.runtime` and `org.eclipse.osgi` helps resolve the re-export package issue and ensures the local JAR package takes precedence.

**Comment by mickaelistria** [comment]
Following up on this, it seems like for this case, we have the following options to do things proper from a OSGi perspective:
1. Make org.eclipse.core.runtime *not* reexport org.eclipse.osgi bundle requirement, or
2. Replace `Require-Bundle: org.eclipse.core.runtime;bundle-version="[3.29,4.0)"` in org.eclipse.jdt.core by some Import-Package

But both have serious drawbacks that make them most likely not desired:
Proposal 1. may break tons of clients (mostly those using the `NLS` class from the org.eclipse.osgi bundle by taking it as transitively provided by org.eclipse core.runtime); and proposal 2. looses the versioning which is very useful for API Tools (mainly because the Export-Package are not versioned themselves in their providers).

So I'm wondering, for this niche use-case, is there a good strategy (such as a system bundle, or some lower-level OSGi hooks) to override how the dependency are being read by Equinox? If so, we could maybe try to just override how the MANIFEST.MF for org.eclipse.jdt.core is read, and just do the replacement we want dynamically, without needing upstream projects to change and thus keeping decent compatibility with those.
What do you think?

**Reply by tjwatson** [reply]
Option 3.  Stop having `org.eclipse.core.runtime` re-export `org.eclipse.osgi`.  Instead have it export all of its APIs only for require-bundle users of `org.eclipse.core.runtime`, but still have `org.eclipse.core.runtime` require-bundle on `org.eclipse.osgi`.  For example, by having it add all the exported packages of `org.eclipse.osgi` but add a mandatory directive like this:

```
Export-Package: org.eclipse.osgi.util; mandatory:="never-import-this-package"
```
This makes `org.eclipse.osgi.util` available as a transitive dependency to anyone that requires `org.eclipse.core.runtime`.  But it will remove all the packages from the JVM that the framework discovers at runtime.

**Reply by merks** [reply]
Note that the package export can be versioned and here it is versioned, though I'm doubtful it's property maintained:

![image](https://github.com/user-attachments/assets/510c83d7-b9bf-4e33-9c06-3fbc68042987)

**Reply by laeubi** [reply]
Actually you can import package with a bundle-version see for example here:

https://stackoverflow.com/questions/36434396/osgi-import-package-version-vs-bundle-version-whats-the-difference

**Reply by merks** [reply]
Yes one can specify bundle-version on the package, but p2 doesn't represent that information so at install time there will be no constraint, only at OSGI runtime.  Also, given the package comes from multiple bundles, which (of the three) bundle's version does one specify?

**Reply by laeubi** [reply]
You can additionally specify `split=common` or `bundle-name` if you want :-)

**Reply by laeubi** [reply]
> Option 3. Stop having org.eclipse.core.runtime re-export org.eclipse.osgi. Instead have it export all of its APIs only for require-bundle users 

@tjwatson I think this is maybe a good compromise, it is not exactly the same, because actually it means it also exports **all system packages** (e.g. javax, org.w3c and alike and specifically the compiler packages here) but as Eclipse runs with boot delegation by default it might not be a big issue.

**Reply by tjwatson** [reply]
I think we should consider making a breaking change and risk breaking require-bundle users that are depending on this nonsensical re-export giving them access to JVM packages.

**Reply by merks** [reply]
Really, there's a bundle-name too?  You learn something new every day, equinox.common also versions that package.

![image](https://github.com/user-attachments/assets/0602e3ea-9800-47cd-ba87-eded3d0cad4c)

Doesn't that eliminate the drawback of option 2 without requiring anything else to change?

**Reply by laeubi** [reply]
> I think we should consider making a breaking change and risk breaking require-bundle users that are depending on this nonsensical re-export giving them access to JVM packages.

I'm all for that, do you like to propose a PR it seems a good time before M1 to do that.

**Reply by laeubi** [reply]
> Really, there's a bundle-name too?

You can find all attributes of the bundle namespace here:
https://docs.osgi.org/specification/osgi.core/8.0.0/framework.namespaces.html#framework.namespaces.osgi.wiring.bundle

and for packages:
https://docs.osgi.org/specification/osgi.core/8.0.0/framework.namespaces.html#framework.namespaces.osgi.wiring.package

**Reply by mickaelistria** [reply]
Indeed, those extra attributes were also unknown to me and I believe that replacing `Require-Bundle` by some `Import-Package: ...;bundle-symbolic-name="org.eclipse.core.runtime";bundle-version="[3.29.0,4.0.0)"` seems to be the shortest path to success for our case. If it works, we'll try to push such change to JDT upstream.
But if someone wants to fix Platform well to remove the reexport once and for all, that would be great too. I just don't think our JDT/Javac-crew will handle it at the moment; but we'd be very pleased to freely benefit from someone else's effort :P

**Reply by tjwatson** [reply]
> > I think we should consider making a breaking change and risk breaking require-bundle users that are depending on this nonsensical re-export giving them access to JVM packages.
> 
> I'm all for that, do you like to propose a PR it seems a good time before M1 to do that.

I think some discussion and a plan needs to happen before jumping into a breaking change.  An announcement to cross-project and the community would be needed IMO before charging ahead.  I said we should consider doing this, not do it immediately.
